### DISCUSSION TITLE
The UI lib using css-in-js cannot render correctly

### DISCUSSION THREAD

**Discussion by qianjiachun** [remix-run/remix#4666] [Q&A] [answered]
Antd v5.0.0 break changes with `css-in-js`. And It no longer needs to import `.css` files.  => [https://ant.design/docs/react/migration-v5](https://ant.design/docs/react/migration-v5)

Then I import it in remix:
1. create a basic remix project.
2. `npm install antd --save`
3. export links in root
```
// app\root.tsx
import stylesAntd from "antd/dist/reset.css"
export const links: LinksFunction = () => [
  { rel: "stylesheet", href: stylesAntd },
];
```
4. set a `Button` in page
```
// app\routes\index.tsx
import { Button } from "antd";

export default function Index() {
  return (
    <Button type="primary">Primary Button</Button>
  );
}
```
5. npm run dev
6. Button not rendered correctly
![image](https://user-images.githubusercontent.com/43806319/203190954-32915331-d23d-436e-ab40-e1fae903f26a.png)

Did I do something wrong?

**Comment by LiHaoGit** [comment, marked as answer]
https://ant.design/docs/react/customize-theme#server-side-render-ssr
https://remix.run/docs/en/v1/guides/styling#css-in-js-libraries

app\root.tsx
```tsx
export default function App() {
  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
        {typeof document === "undefined" ? "__STYLES__" : null}
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```


app\entry.server.tsx
```tsx
import { createCache, StyleProvider, extractStyle } from "@ant-design/cssinjs";

function handleBrowserRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const cache = createCache();
  let markup = renderToString(
    <StyleProvider cache={cache}>
      <RemixServer context={remixContext} url={request.url} />
    </StyleProvider>
  );
  const styleText = extractStyle(cache);
  
  responseHeaders.set("Content-Type", "text/html");
  markup = markup.replace("__STYLES__", styleText);
  return new Response(`<!DOCTYPE html> ${markup}`, {
    status: responseStatusCode,
    headers: responseHeaders,
  });
}
```
It's working!
![image](https://user-images.githubusercontent.com/7036191/203463846-1abd80bf-9720-49b6-95d7-85114d285722.png)

**Reply by qianjiachun** [reply]
worked! Hearty thanks.

**Reply by qianjiachun** [reply]
hmmm, I have two questions:
1. After doing this, console shows `Warning: Extra attributes from the server: style`.
2. The component style will be loaded after the page is loaded, which causes the jump of the elements.

**Reply by LiHaoGit** [reply]
我更新了代码，你在试一下。
分别参考了    
https://remix.run/docs/en/v1/guides/styling#css-in-js-libraries       
https://ant.design/docs/react/customize-theme#server-side-render-ssr

**Reply by qianjiachun** [reply]
感谢帮助。
1. `Warning: Extra attributes from the server: style` 的问题在build之后就没有啦😂
2. 页面元素跳动的问题稍微好了一些，但是还是存在，可能是我覆盖了一些antd的样式？或者我用了tailwind？

如果需要提供代码的话我去开一个repo

**Reply by qianjiachun** [reply]
> 我更新了代码，你在试一下。 分别参考了 https://remix.run/docs/en/v1/guides/styling#css-in-js-libraries https://ant.design/docs/react/customize-theme#server-side-render-ssr

我的项目: [https://github.com/qianjiachun/csgo-skins-recorder-fe](https://github.com/qianjiachun/csgo-skins-recorder-fe)
1. `npm install`
2. `npm run dev`
3. 可以看到刷新页面后，一些样式会延迟生效😂如果您方便的话可以帮忙看一下

**Reply by machour** [reply]
Could you guys keep the conversation in english so we, and future readers, can follow along ? 🙏🏼

**Reply by qianjiachun** [reply]
> Could you guys keep the conversation in english so we, and future readers, can follow along ? 🙏🏼

Okay Okay !

**Comment by qianjiachun** [comment]
The problem is solved, and the method is consistent with answer. 
The problem with "element jump" is that I wrote an inappropriate style. It has nothing to do with the lib. 

Thanks again for your help!
